Zend/tests/constexpr/constant_expressions_cast.phpt

```diff
@@ -0,0 +1,51 @@
+--TEST--
+Constant expressions with cast
+--FILE--
+<?php
+class X {
+    public int $foo = 3;
+}
+
+const T1 = (int) 0.3;
+const T2 = (bool) 0.3;
+const T3 = (string) [];
+const T4 = (object) ["a" => 1];
+const T5 = (float) 5;
+const T6 = (array) "";
+const T7 = (array) var_dump(...);
```

Comment:
Test with warning like https://3v4l.org/6ugOa should be added.

IDK if exception is possible, if so, it should be tested as well.

Reply:
Exceptions are at least possible with `(string) $nonStringable`, there might be others.

Reply:
I think the point is that the `IS_STRING` branch should set `ret` to `FAILURE` if the cast fails. I realized that we don't handle this incorrectly for destructors too. Maybe we should add an `if (EG(exception)) { ret = FAILURE; }` right before the return. That said, this seems to not cause any issues because neither constructors nor destructors are called from `zend_ast_evaluate_inner()` when `EG(exception)` is set. So maybe it's ok, `zend_ast_evaluate_inner()` will just do too much work when an exception occurs early.